src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,708 @@
+  // https://wiki.squid-cache.org/SquidFaq/InnerWorkings#What_is_a_half-closed_filedescriptor.3F
+  httpAllowHalfOpen = false;
+  timeout = 0;
+  maxHeadersCount: number | null = null;
+  maxRequestsPerSocket = 0;
+
+  requestTimeout: number = 0;
+  headersTimeout: number = 0;
+  requireHostHeader: boolean = false;
+  joinDuplicateHeaders: boolean = false;
+  rejectNonStandardBodyWrites: boolean = false;
+  keepAliveTimeout: number = 5_000;
+  port?: number;
+
+  constructor(options?: ServerOptions, requestListener?: RequestListener) {
+    super();
+
+    if (options != null) {
+      storeHTTPOptions.call(this, options);
+    }
+
+    if (typeof options === 'function') {
+      requestListener = options;
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this.on('listening', setupConnectionsTracking);
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(): this {
+    httpServerPreClose(this);
+    if (this.port) {
+      portMapper.delete(this.port);
+    }
+    return this;
+  }
+
+  closeAllConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  closeIdleConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  setTimeout(
+    msecs?: number | ((socket: Socket) => void),
+    callback?: (socket: Socket) => void
+  ): this {
+    if (typeof msecs === 'function') {
+      callback = msecs;
+      msecs = undefined;
+    } else if (typeof msecs === 'number') {
+      this.timeout = msecs;
+    }
+
+    if (typeof callback === 'function') {
+      this.once('timeout', callback);
+    }
+    return this;
+  }
+
+  async #onRequest(request: Request): Promise<Response> {
+    const { incoming, response } = this.#toReqRes(request);
+    this.emit('connection', this, incoming);
+    this.emit('request', incoming, response);
+    return getServerResponseFetchResponse(response);
+  }
+
+  #toReqRes(request: Request): {
+    incoming: IncomingMessage;
+    response: ServerResponse;
+  } {
+    const incoming = new IncomingMessage();
+    const reqUrl = new URL(request.url);
+    incoming.url = reqUrl.pathname + reqUrl.search;
+
+    const headers = [];
+    for (const [key, value] of request.headers) {
+      if (key === 'host') {
+        // By default fetch implementation will join "host" header values with a comma.
+        // But in order to be node.js compatible, we need to select the first if possible.
+        headers.push(key, value.split(', ').at(0) as string);
+      } else {
+        headers.push(key, value);
+      }
+    }
+    incoming._addHeaderLines(headers, headers.length);
+
+    incoming.method = request.method;
+    incoming._stream = request.body;
```

Comment:
Looking over `IncomingStream`, it looks like this is currently not used at all? So the request payload is not consumed at all yet?

Reply:
Amazing eye! Yes, currently working on it.